add Topological Optimization code

MohamedKISSI opened the pull request:
Integration of the code for topological simplification based on persistence diagram optimization.

MohamedKISSI committed:
add Topological Optimization code

MohamedKISSI committed:
clang-format the code

julien-tierny commented:
Hi Mohamed,
thanks a lot for this PR.

At this point, the PR doesn't pass the continuous integration (CI), which stopped on a third-party dependency issue (here, torch). As you will read in this documentation (https://github.com/topology-tool-kit/ttk/wiki/Guidelines-for-Developing-a-New-TTK-Module), TTK must build and run even when third-party dependencies haven't been found.

This means that your code should build and run, even if torch is not installed. In particular, if the optimization backend selected by the user was Adam and that Torch was not installed, you should display a warning (`printWrn`) explaining that Adam cannot be used because Torch hasn't been found and that the code will now default to direct gradient descent.

To know in your C++ code if Torch has been found or not, use the preprocessor instruction  `#ifdef TTK_ENABLE_TORCH` (see https://github.com/topology-tool-kit/ttk/blob/dev/core/base/mergeTreeAutoencoder/MergeTreeAutoencoder.h for an example). 

Please proceed to this change and fix any issue until the CI succeeds.
Thanks

MohamedKISSI committed:
Fixing torch dependency issue

MohamedKISSI committed:
Give a default value to the constraintDiagram variable so that the ttk examples work

MohamedKISSI committed:
set constraintDiagram variable to const

MohamedKISSI committed:
We remove the default initialization for certain variables

MohamedKISSI committed:
fix odr violations

MohamedKISSI committed:
fix odr violations getCoordinatesInformations

julien-tierny committed:
restored original gui for existing backends

julien-tierny committed:
progress on ui shuffling

julien-tierny committed:
work on solver gui

julien-tierny committed:
updated optimization gui

MohamedKISSI committed:
Set default variable values to match XML file and updated PyTorch warning messages.

julien-tierny committed:
Merge branch 'dev' of https://github.com/topology-tool-kit/ttk into topologicalOptimization

julien-tierny committed:
updated adam unavailability message

julien-tierny committed:
topological optimization backend refactoring

julien-tierny committed:
Reformatting console outputs

julien-tierny committed:
trying to debug windows ci

julien-tierny committed:
listing conda installation

julien-tierny committed:
added missing files

julien-tierny committed:
listing more conda install

julien-tierny committed:
exploring conda install

julien-tierny committed:
try to add libboost-devel

julien-tierny commented:
- Code tested on a simple example (with and without Torch support, with multiple options)
- PV GUI edited
- base code refactored
- console messages edited

julien-tierny committed:
searching for conda's boost cmake files

julien-tierny committed:
simpler windows CI fix

julien-tierny committed:
fix refactoring

julien-tierny committed:
removed unnecessary headers

julien-tierny committed:
updated var names

julien-tierny committed:
remvod comments

julien-tierny committed:
fixed vtk layer

julien-tierny committed:
used proper indices, removed unused functions

julien-tierny commented on core/base/topologicalOptimization/TopologicalOptimization.h:
```diff
@@ -0,0 +1,1674 @@
+        -   neighborsIndices : vector which contains the neighboring vertices of
+      vertex i
+    */
+    template <typename triangulationType>
```
Please remove that function (see my comments below when it is used).

julien-tierny commented on core/base/topologicalOptimization/TopologicalOptimization.h:
```diff
@@ -0,0 +1,1674 @@
+          needUpdate[indice] = true;
+          // Find all the neighbors of the vertex
+          std::vector<SimplexId> neighborsIndices;
+          getNeighborsIndices(triangulation, indice, neighborsIndices);
```
please remove that call.

julien-tierny commented on core/base/topologicalOptimization/TopologicalOptimization.h:
```diff
@@ -0,0 +1,1674 @@
+          std::vector<SimplexId> neighborsIndices;
+          getNeighborsIndices(triangulation, indice, neighborsIndices);
+
+          for(SimplexId neighborsIndice : neighborsIndices) {
```
replace that loop by:
```
int vertexNumber = triangulation->getVertexNeighborNumber(index);
for(int i = 0; i < vertexNumber; i++){
   int vertexNeighborId = -1;
   triangulation->getVertexNeighbor(index, i, vertexNeighborId);
   needUpdate[vertexNeighborId] = true;
}
```

julien-tierny committed:
updated var names

julien-tierny commented on core/base/topologicalOptimization/TopologicalOptimization.h:
```diff
@@ -0,0 +1,1674 @@
+  vertex i
+*/
+template <typename triangulationType>
+int ttk::TopologicalOptimization::getNeighborsIndices(
```
to remove.

julien-tierny committed:
typo fix

julien-tierny commented on core/base/topologicalOptimization/TopologicalOptimization.h:
```diff
@@ -0,0 +1,1674 @@
+  //=======================
+  //    Copy input data
+  //=======================
+  std::vector<double> dataVector(vertexNumber_);
```
Why is the data temporarily converted to double for the computation?
Is that a requirement from Torch?

julien-tierny reviewed (commented):
Hi Mohamed, 
thanks again for this PR.
I did quite some modifications to it.

I left a question in my review as well as a request for removing yet another function.
thanks!

julien-tierny committed:
updated paper references

MohamedKISSI committed:
Removed getNeighborsIndices function and updated dataVector type to dataType.

MohamedKISSI commented:
> Hi Mohamed, thanks again for this PR. I did quite some modifications to it.
> 
> I left a question in my review as well as a request for removing yet another function. thanks!

Hi Julien, 

Thank you for your feedback. It's duly noted; I have removed the getNeighborsIndices function and changed the type of the dataVector variable to dataType.
 Thanks

julien-tierny committed:
Merge branch 'dev' of https://github.com/topology-tool-kit/ttk into topologicalOptimization

julien-tierny committed:
clang-format

julien-tierny committed:
removed unnecessary headers

julien-tierny commented:
Hi Mohamed,

thanks a lot for your edits.
We need to discuss a couple of things though:
- necessity of double format for the data vector (see `kFloat64`), as a first effort, please use double for dataVectors and make all the computation in double.
- check consistency between `kFloat64` and `kDouble`
- multiple unnecessary copies of the ouput data `dataVector` `smoothedScalars`

julien-tierny commented:
Moreover, the input data must be first normalized in a pre-process, and "de-normalized" in a post-process.

julien-tierny commented:
Also, please report a status message every ten iterations (and not every iterations). Thanks!
EDIT: make the value 10 editable from the GUI (with 10 being a default value)

MohamedKISSI committed:
Use double format for dataVectors and computations, ensure consistency between kFloat64 and kDouble, remove unnecessary copies of smoothedScalars. Normalize input data in pre-process and de-normalize in post-process. Report status message every k iterations.

julien-tierny commented:
hi mohamed,
thanks for your quick fix.

it seems that your last commit is breaking things though.

please consider the statefile I sent you in slack, in the TopologicalSimplification node, switch from LTS to TopologicalOptimization and you'll see that the direct gradient descent method generates incorrect results (the adam method looks OK).

MohamedKISSI committed:
Solving problems by normalizing the values ​​of the critical points of the target diagram.

julien-tierny commented:
OK, looks good to me. Thanks a lot!
